File: api/generate-pdf.js
const { PDFDocument, rgb, StandardFonts, PDFName, PDFBool } = require('pdf-lib');
const fs = require('fs');
const path = require('path');

module.exports.config = {
  api: { bodyParser: { sizeLimit: '2mb' } }
};

const CHECKBOXES = [
  ['Special diet',    39.4, 423.9,  69.6, 424.1],
  ['Allergies',       39.4, 399.7,  69.1, 400.1],
  ['Self Admin',      39.9, 352.1,  69.7, 352.1],
  ['Chronic illness', 39.6, 288.1,  69.1, 288.2],
  ['Surgery',         39.5, 312.1,  69.2, 312.0],
];

async function handler(req, res) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

  try {
    let data = req.body;
    if (typeof data === 'string') {
      try { data = JSON.parse(data); } catch(e) { return res.status(400).json({ error: 'Invalid JSON' }); }
    }
    if (!data) return res.status(400).json({ error: 'Empty body' });

    // Find the blank PDF
    const possiblePaths = [
      path.join(process.cwd(), 'Permission_Blank.pdf'),
      path.join(process.cwd(), 'public', 'Permission_Blank.pdf'),
      path.join(__dirname, '..', 'Permission_Blank.pdf'),
    ];
    let pdfBytes = null;
    for (const p of possiblePaths) {
      try { if (fs.existsSync(p)) { pdfBytes = fs.readFileSync(p); break; } } catch(e) {}
    }
    if (!pdfBytes) return res.status(500).json({
      error: 'Permission_Blank.pdf not found',
      cwd: process.cwd(),
      tried: possiblePaths,
      ls: (() => { try { return fs.readdirSync(process.cwd()); } catch(e) { return []; } })()
    });

    const pdfDoc = await PDFDocument.load(pdfBytes);
    const form = pdfDoc.getForm();
    const page = pdfDoc.getPages()[0];

    // Set text fields — no updateAppearances, just set values
    // NeedAppearances tells the viewer to render them using its own fonts
    function setField(name, value) {
      try {
        if (value === null || value === undefined || value === '') return;
        form.getTextField(name).setText(String(value));
      } catch(e) {}
    }

    // Event details
    setField('Event',                                   data.evName);
    setField('Dates of event',                          data.evDates);
    setField('Event description',                       data.evDesc);
    setField('Ward',                                    data.evWard);
    setField('Stake',                                   data.evStake);
    setField('Event or activity leader',                'David Wheat (Bishop)');
    setField('Event or activity leaders phone number',  '210-601-2913');
    setField('Event or activity leaders email',         'creamof@gmail.com');

    // Participant
    setField('Participant',                             data.name);
    setField('Date of birth',                           data.dob);
    setField('Age',                                     data.age);
    setField('Telephone number',                        data.tel);
    setField('Address',                                 data.addr);
    setField('City',                                    data.city);
    setField('State or Province',                       data.state);
    setField('Emergency contact parent or guardian',    data.ec);
    setField('Primary phone_1',                         data.ec1);
    setField('Secondary phone_1',                       data.ec2);
    setField('List of Medications',                     data.meds);
    setField('diet explanation',                        data.diet);
    setField('Allergy explanation',                     data.allergy);
    setField('illness explanation',                     data.chronic);
    setField('If yes please explain_2',                 data.surgery);
    setField('Other limitations',                       data.limits);
    setField('Special needs',                           data.other);

    // Tell the viewer to render all field appearances from their values
    // This is the standard way — viewer uses its own fonts, works everywhere
    try {
      const acroForm = pdfDoc.catalog.lookup(PDFName.of('AcroForm'));
      if (acroForm) acroForm.set(PDFName.of('NeedAppearances'), PDFBool.True);
    } catch(e) {}

    // Draw X marks directly on the page for checkboxes
    const choices = {
      'Special diet':    data.diet_yn    || 'no',
      'Allergies':       data.allergy_yn || 'no',
      'Self Admin':      data.selfadmin  || 'yes',
      'Chronic illness': data.chronic_yn || 'no',
      'Surgery':         data.surgery_yn || 'no',
    };
    const s = 3.2;
    const black = rgb(0, 0, 0);
    for (const [name, yesX, yesY, noX, noY] of CHECKBOXES) {
      const cx = choices[name] === 'yes' ? yesX : noX;
      const cy = choices[name] === 'yes' ? yesY : noY;
      page.drawLine({ start: { x: cx-s, y: cy-s }, end: { x: cx+s, y: cy+s }, thickness: 1.5, color: black });
      page.drawLine({ start: { x: cx+s, y: cy-s }, end: { x: cx-s, y: cy+s }, thickness: 1.5, color: black });
    }

    // Save with useObjectStreams: false — produces a traditional xref table
    // compatible with ALL PDF viewers including old macOS Preview versions
    const result = await pdfDoc.save({ useObjectStreams: false });

    const safeName = (data.name || 'form').replace(/[^a-z0-9]/gi, '_');
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="PermissionForm_${safeName}.pdf"`);
    res.send(Buffer.from(result));

  } catch(err) {
    console.error('PDF error:', err);
    res.status(500).json({ error: err.message, stack: err.stack });
  }
}

module.exports = handler;

File: api/generate-pdf.py
from http.server import BaseHTTPRequestHandler
import json, os, io
from reportlab.pdfgen import canvas
from reportlab.lib.colors import black
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from pypdf import PdfReader, PdfWriter
from pypdf.generic import NameObject, create_string_object

CHECKBOXES = [
    ('Special diet',    39.4, 423.9,  69.6, 424.1),
    ('Allergies',       39.4, 399.7,  69.1, 400.1),
    ('Self Admin',      39.9, 352.1,  69.7, 352.1),
    ('Chronic illness', 39.6, 288.1,  69.1, 288.2),
    ('Surgery',         39.5, 312.1,  69.2, 312.0),
]

def get_pdf_path():
    base = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    for p in [
        os.path.join(base, 'Permission_Blank.pdf'),
        os.path.join(base, 'public', 'Permission_Blank.pdf'),
        os.path.join(os.getcwd(), 'Permission_Blank.pdf'),
    ]:
        if os.path.exists(p):
            return p
    raise FileNotFoundError(f'Permission_Blank.pdf not found. base={base} cwd={os.getcwd()}')

def fill_pdf(data):
    pdf_path = get_pdf_path()

    # Step 1: fill text fields using pypdf
    r = PdfReader(pdf_path)
    w = PdfWriter()
    w.append(r)

    TEXT_FIELDS = {
        'Event':                                    data.get('evName', ''),
        'Dates of event':                           data.get('evDates', ''),
        'Event description':                        data.get('evDesc', ''),
        'Ward':                                     data.get('evWard', ''),
        'Stake':                                    data.get('evStake', ''),
        'Event or activity leader':                 data.get('evLeader', 'David Wheat'),
        'Event or activity leaders phone number':   data.get('evPhone', '210-601-2913'),
        'Event or activity leaders email':          data.get('evEmail', 'creamof@gmail.com'),
        'Participant':                              data.get('name', ''),
        'Date of birth':                            data.get('dob', ''),
        'Age':                                      data.get('age', ''),
        'Telephone number':                         data.get('tel', ''),
        'Address':                                  data.get('addr', ''),
        'City':                                     data.get('city', ''),
        'State or Province':                        data.get('state', ''),
        'Emergency contact parent or guardian':     data.get('ec', ''),
        'Primary phone_1':                          data.get('ec1', ''),
        'Secondary phone_1':                        data.get('ec2', ''),
        'List of Medications':                      data.get('meds', ''),
        'diet explanation':                         data.get('diet', ''),
        'Allergy explanation':                      data.get('allergy', ''),
        'illness explanation':                      data.get('chronic', ''),
        'If yes please explain_2':                  data.get('surgery', ''),
        'Other limitations':                        data.get('limits', ''),
        'Special needs':                            data.get('other', ''),
    }
    w.update_page_form_field_values(w.pages[0], TEXT_FIELDS, auto_regenerate=False)

    # Set NeedAppearances so viewer renders field values
    if '/AcroForm' in w._root_object:
        w._root_object['/AcroForm'].update({
            NameObject('/NeedAppearances'): create_string_object('true')
        })

    # Write filled PDF to buffer
    buf1 = io.BytesIO()
    w.write(buf1)
    buf1.seek(0)

    # Step 2: draw X marks as overlay using reportlab
    choices = {
        'Special diet':    data.get('diet_yn', 'no'),
        'Allergies':       data.get('allergy_yn', 'no'),
        'Self Admin':      data.get('selfadmin', 'yes'),
        'Chronic illness': data.get('chronic_yn', 'no'),
        'Surgery':         data.get('surgery_yn', 'no'),
    }
    packet = io.BytesIO()
    c = canvas.Canvas(packet, pagesize=(612, 792))
    c.setStrokeColor(black)
    c.setLineWidth(1.3)
    s = 3.2
    for (name, yesX, yesY, noX, noY) in CHECKBOXES:
        cx = yesX if choices.get(name) == 'yes' else noX
        cy = yesY if choices.get(name) == 'yes' else noY
        c.line(cx-s, cy-s, cx+s, cy+s)
        c.line(cx+s, cy-s, cx-s, cy+s)
    c.save()
    packet.seek(0)

    # Step 3: merge overlay onto filled PDF
    overlay_reader = PdfReader(packet)
    filled_reader = PdfReader(buf1)
    final_writer = PdfWriter()
    final_writer.append(filled_reader)
    final_writer.pages[0].merge_page(overlay_reader.pages[0])

    out = io.BytesIO()
    final_writer.write(out)
    return out.getvalue()


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(length)
            data = json.loads(body)
            pdf_bytes = fill_pdf(data)
            name = (data.get('name', 'form') or 'form').replace(' ', '_')
            self.send_response(200)
            self.send_header('Content-Type', 'application/pdf')
            self.send_header('Content-Disposition',
                f'attachment; filename="PermissionForm_{name}.pdf"')
            self.send_header('Content-Length', str(len(pdf_bytes)))
            self.end_headers()
            self.wfile.write(pdf_bytes)
        except Exception as e:
            import traceback
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'error': str(e),
                'trace': traceback.format_exc()
            }).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
